datumaro/components/dataset.py

```diff
@@ -1449,6 +1449,32 @@ def detect(path: str, *, env: Optional[Environment] = None, depth: int = 2) -> s
         else:
             return matches[0]
 
+    @staticmethod
+    def detect_version(
+        path: str, *, env: Optional[Environment] = None, depth: Optional[int] = None
+    ) -> Dict[str, str]:
+        """
+        Detect the datumaro version of dataset at exported time
+
+        Args:
+            path: path to dataset
+            env: A plugin collection. If not set, the built-in plugins are used
+        """
+        detect_kwargs = {}
+        if env is not None:
+            detect_kwargs["env"] = env
+        if depth is not None:
+            detect_kwargs["depth"] = depth
+
+        format_name = Dataset.detect(path, **detect_kwargs)
+
+        if env is None:
+            env = Environment()
+
+        version = env.detect_version(path=path, format_name=format_name)
+
+        return version
+
```

Comment:
I think that dataset versioning is only affected to Datumaro format so that we don't need to expose this to `Dataset` level. I guess this versioning can be dumpped to the `Dataset.info` dictionary, e.g., `key("dm_format_version")-value("1.0")`.

Reply:
Or if you really want to add this to `Dataset`, I think it would be better to create a member variable, e.g., `Dataset._format_version` and make a setter for this variable in the dataset extractor.

Reply:
For `backward compatibility', I think that we should be able to get version string from the dataset path before creating Dataset object. Thus, I think that member variable or using a `infos()` are not good place to store version.

Reply:
My opinion is that the one who create `Dataset` is each `DatasetBase` (`Extractor` in the past, e.g., `DatumaroBase`) and it can decide that the current Datumaro has compatibility with the dataset version exactly when it tries to import the dataset from the file. For example,

```python
class DatumaroBase(DatasetBase):
    def __init__(self, path):
    version = parse_version(path)
    if not check_version_compatibility():
         raise DatasetImporterror(f"version {version} is not compatible. Please use the future version of Datumaro.")
    
    if version == "1.0"
         self._data = self.reader_v1_0()
    elif version == "1.1"
         self._data = self.reader_v1_1()
     ...
```

Reply:
I agree that it's not necessary to expose the version to `Dataset` level.

Maybe some alternative could be generating a new file to store the overall meta-info of the Datumaro dataset not in the json for annotation.
@wonjuleee @vinnamkim Please share your opinion.

Reply:
> My opinion is that the one who create `Dataset` is each `DatasetBase` (`Extractor` in the past, e.g., `DatumaroBase`) and it can decide that the current Datumaro has compatibility with the dataset version exactly when it tries to import the dataset from the file. For example,
> 
> ```python
> class DatumaroBase(DatasetBase):
>     def __init__(self, path):
>     version = parse_version(path)
>     if not check_version_compatibility():
>          raise DatasetImporterror(f"version {version} is not compatible. Please use the future version of Datumaro.")
>     
>     if version == "1.0":
>          self._data = self.reader_v1_0()
>     elif version == "1.1":
>          self._data = self.reader_v1_1()
>      ...
> ```

Yes, I guess that at this point, we don't fully generate Dataset object thus we should avoid storing the exported version in the Dataset obejct.

Reply:
I discuss about this topic with @vinnamkim on the 1:1 chat.

I agree that the scope of versioning in the PR shouldn't affect all dataset types, only datumaro.
Thus I'll limit the changes on the datumaro plugin side.

About where to place the version string,
This should be volatility (removed after reading), I'll make a new field in json rather than using existing field, infos, because they are loaded to Dataset object while reading.